gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/auth/user/impl/UserAuthenticationManagerImpl.java

```diff
@@ -193,6 +182,38 @@
                 });
     }
 
+    private List<ApplicationIdentityProvider> getApplicationIdentityProviders(Client client, Authentication authentication) {
+        // Get identity providers associated to a client
+        // For each idp, try to authenticate a user
+        // Try to authenticate while the user can not be authenticated
+        // If user can't be authenticated, send an exception
+
+        // Skip external identity provider for authentication with credentials.
+        if (isNull(client.getIdentityProviders())) {
+            return List.of();
+        }
+        return client.getIdentityProviders().stream().filter(appIdp -> {
+                    var identityProvider = identityProviderManager.getIdentityProvider(appIdp.getIdentity());
+                    return nonNull(identityProvider) && !identityProvider.isExternal();
+                })
+                .collect(toList());
+    }
+
+    private boolean selectionRuleMatches(String rule, Authentication authentication) {
+        try {
+            // We keep the idp if the rule is not present to keep the same behaviour
+            // The priority will define the order of the identity providers to match the rule first
+            if (Strings.isNullOrEmpty(rule) || rule.isBlank()) {
+                return true;
+            }
+            var templateEngine = authentication.getContext().getTemplateEngine();
+            return templateEngine != null && templateEngine.getValue(rule.trim(), Boolean.class);
+        } catch (Exception e) {
```

Comment:
Perhaps we should only attempt to catch `ParseException` here, even if it is only to log?

Reply:
I believe that if there is any exception happening (there could be unhinged NPEs), the business case should return false so that it does not disturb users in the login flow

However, logging the exception does not cost much